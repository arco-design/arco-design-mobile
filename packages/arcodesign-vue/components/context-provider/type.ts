export interface ContextInstallOptions {
    /**
     * 组件名称前缀
     * @en Component prefix
     * @default "Arco"
     */
    componentPrefix?: string;
}

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
}
