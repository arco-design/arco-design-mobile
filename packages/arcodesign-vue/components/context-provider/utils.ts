import { InjectionKey, inject } from 'vue';
import { ContextInstallOptions, GlobalContextParams } from './type';

export const defaultInstallOptions: Required<ContextInstallOptions> = {
    componentPrefix: 'Arco',
};

export const defaultContext: Required<GlobalContextParams> = {
    prefixCls: 'arco',
};

export const configProviderInjectionKey: InjectionKey<GlobalContextParams> =
    Symbol('ArcoContextProvider');

export function getComponentPrefix(options?: ContextInstallOptions) {
    return options?.componentPrefix || defaultInstallOptions.componentPrefix;
}

export function getPrefixCls(componentName?: string) {
    const contextProvider = inject(configProviderInjectionKey, defaultContext);
    const prefix = contextProvider.prefixCls;
    if (componentName) {
        return `${prefix}-${componentName}`;
    }
    return prefix;
}
