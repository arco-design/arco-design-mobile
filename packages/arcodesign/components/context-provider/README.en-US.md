### Others

# ContextProvider 

Global data control component, used to replace global data.

======

> Props

|Property|Description|Type|DefaultValue|
|----------|-------------|------|------|
|prefixCls|Component classname prefix|string|"arco"|
|system|Manually control the current system, and the incoming value will be used directly after being passed in\. It is applicable when the initial value of the system needs to be specified in the ssr scenario\.|"" \| "pc" \| "android" \| "ios"|""|
|useDarkMode|Whether to use dark mode|boolean|false|
|isDarkMode|Whether it is in dark mode|boolean|false|
|theme|Theme variable\. The css variable will be replaced online after input\. The less variable needs to be set|Record\<string, string\>|-|
|locale|Internationalized language configuration|ILocale|-|
|useRtl|Whether to use rtl|boolean|false|

> ILocale

|Property|Description|Type|
|----------|-------------|------|
|locale|Language Type|string|
|LoadMore|-|\{ loadMoreText: string; loadingText: string; prepareText: string; noDataText: string; failLoadText: string; prepareScrollText: string; prepareClickText: string; \}|
|Picker|-|\{ okText: string; cancelText: string; \}|
|Tag|-|\{ addTag: string; \}|
|Dialog|-|\{ okText: string; cancelText: string; \}|
|SwipeLoad|-|\{ normalText: string; activeText: string; \}|
|PullRefresh|-|\{ loadingText: string; pullingText: string; finishText: string; loosingText: string; \}|
|DropdownMenu|-|\{ select: string; \}|
|Pagination|-|\{ previousPage: string; nextPage: string; \}|
|Image|-|\{ loadError: string; \}|
|ImagePicker|-|\{ loadError: string; \}|
|SearchBar|-|\{ placeholder: string; cancelBtn: string; \}|
|Stepper|-|\{ minusButtonName: string; addButtonName: string; \}|
|Keyboard|-|\{ confirm: string; \}|
|Form|-|\{ required: string; type: \{ email: string; url: string; string: string; number: string; array: string; object: string; boolean: string; \}; number: \{ min: string; max: string; equal: string; range: string; positive: string; negative: string; \}; string: \{ \.\.\.; \}; array: \{ \.\.\.; \}; object: \{ \.\.\.; \}; boolean: \{ \.\.\.; \}; \}|
