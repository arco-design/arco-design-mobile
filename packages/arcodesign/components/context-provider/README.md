### 其他

# 全局配置 ContextProvider

全局数据控制组件，用于替换全局数据。

======

> 属性/Props

|参数|描述|类型|默认值|
|----------|-------------|------|------|
|prefixCls|组件类名前缀|string|"arco"|
|system|手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用|"" \| "pc" \| "android" \| "ios"|""|
|useDarkMode|是否使用暗黑模式|boolean|false|
|isDarkMode|是否处于暗黑模式，指定后以指定的值为准|boolean|false|
|theme|主题变量，传入后将在线替换css变量，需设置less变量 @use\-css\-vars: 1|Record\<string, string\>|-|
|locale|国际化语言包配置|ILocale|-|
|useRtl|是否使用Rtl模式|boolean|false|

> ILocale

|参数|描述|类型|
|----------|-------------|------|
|locale|语言类型|string|
|LoadMore|-|\{ loadMoreText: string; loadingText: string; prepareText: string; noDataText: string; failLoadText: string; prepareScrollText: string; prepareClickText: string; \}|
|Picker|-|\{ okText: string; cancelText: string; \}|
|Tag|-|\{ addTag: string; \}|
|Dialog|-|\{ okText: string; cancelText: string; \}|
|SwipeLoad|-|\{ normalText: string; activeText: string; \}|
|PullRefresh|-|\{ loadingText: string; pullingText: string; finishText: string; loosingText: string; \}|
|DropdownMenu|-|\{ select: string; \}|
|Pagination|-|\{ previousPage: string; nextPage: string; \}|
|Image|-|\{ loadError: string; \}|
|ImagePicker|-|\{ loadError: string; \}|
|SearchBar|-|\{ placeholder: string; cancelBtn: string; \}|
|Stepper|-|\{ minusButtonName: string; addButtonName: string; \}|
|Keyboard|-|\{ confirm: string; \}|
|Form|-|\{ required: string; type: \{ email: string; url: string; string: string; number: string; array: string; object: string; boolean: string; \}; number: \{ min: string; max: string; equal: string; range: string; positive: string; negative: string; \}; string: \{ \.\.\.; \}; array: \{ \.\.\.; \}; object: \{ \.\.\.; \}; boolean: \{ \.\.\.; \}; \}|
