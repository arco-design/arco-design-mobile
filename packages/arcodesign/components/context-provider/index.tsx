import React, { createContext, useState, useEffect, useCallback } from 'react';
import { addCssRules, removeCssStyleDom, defaultLocale, ILocale } from '@arco-design/mobile-utils';

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
    /**
     * 手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用
     * @en Manually control the current system, and the incoming value will be used directly after being passed in. It is applicable when the initial value of the system needs to be specified in the ssr scenario.
     * @default ""
     */
    system?: 'pc' | 'android' | 'ios' | '';
    /**
     * 是否使用暗黑模式
     * @en Whether to use dark mode
     * @default false
     */
    useDarkMode?: boolean;
    /**
     * 是否处于暗黑模式，指定后以指定的值为准
     * @en Whether it is in dark mode
     * @default false
     */
    isDarkMode?: boolean;
    /**
     * 主题变量，传入后将在线替换css变量，需设置less变量 @use-css-vars: 1
     * @en Theme variable. The css variable will be replaced online after input. The less variable needs to be set @use-css-vars: 1
     */
    theme?: Record<string, string>;
    /**
     * 国际化语言包配置
     * @en Internationalized language configuration
     */
    locale?: ILocale;
    /**
     * 是否使用Rtl模式
     * @en Whether to use rtl
     * @default false
     */
    useRtl?: boolean;
}

export const defaultContext: GlobalContextParams = {
    prefixCls: 'arco',
    system: '',
    useDarkMode: false,
    isDarkMode: false,
    locale: defaultLocale,
    useRtl: false,
};

export const GlobalContext = createContext<GlobalContextParams>(defaultContext);

export interface ContextProviderProps extends GlobalContextParams {
    children: React.ReactNode;
}

export type WithGlobalContext<T> = T & { context?: GlobalContextParams };

/**
 * 全局数据控制组件，用于替换全局数据。
 * @en Global data control component, used to replace global data.
 * @type 其他
 * @type_en Others
 * @name 全局配置
 * @name_en ContextProvider
 */
export default function ContextProvider(props: ContextProviderProps) {
    const {
        children,
        useDarkMode,
        isDarkMode,
        theme,
        locale = defaultLocale,
        ...restProps
    } = props;
    const [isDarkModeState, setIsDarkModeState] = useState<boolean>();
    const changeDarkMode = useCallback(
        (res: MediaQueryListEvent) => setIsDarkModeState(res.matches),
        [],
    );

    useEffect(() => {
        const matchMedia = window.matchMedia('(prefers-color-scheme: dark)');
        if (useDarkMode) {
            const dark = matchMedia.matches;
            setIsDarkModeState(dark);
            if (typeof matchMedia.addEventListener === 'function') {
                matchMedia.addEventListener('change', changeDarkMode);
            } else if (typeof matchMedia.addListener === 'function') {
                matchMedia.addListener(changeDarkMode);
            }
        }
        return () => {
            if (useDarkMode) {
                if (typeof matchMedia.removeEventListener === 'function') {
                    matchMedia.removeEventListener('change', changeDarkMode);
                } else if (typeof matchMedia.removeListener === 'function') {
                    matchMedia.removeListener(changeDarkMode);
                }
            }
        };
    }, [useDarkMode]);

    useEffect(() => {
        if (!theme) {
            return;
        }
        addCssRules('arcoCustomTheme', theme);
        return () => {
            removeCssStyleDom('arcoCustomTheme');
        };
    }, [theme]);

    return (
        <GlobalContext.Provider
            value={{
                ...defaultContext,
                ...restProps,
                isDarkMode: isDarkMode === void 0 ? isDarkModeState : isDarkMode,
                useDarkMode,
                locale: locale || defaultLocale,
            }}
        >
            {children}
        </GlobalContext.Provider>
    );
}

export const ContextLayout = GlobalContext.Consumer;

export function CompWithGlobalContext<P extends JSX.IntrinsicAttributes>(
    Component: React.FunctionComponent<P>,
) {
    return function (props: WithGlobalContext<P>) {
        const { context: propsContext, ...others } = props;
        return (
            <ContextLayout>
                {context => (
                    <ContextProvider {...{ ...context, ...propsContext }}>
                        <Component {...(others as P)} />
                    </ContextProvider>
                )}
            </ContextLayout>
        );
    };
}
