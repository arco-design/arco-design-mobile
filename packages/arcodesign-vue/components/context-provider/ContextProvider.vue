<script lang="ts" setup>
import { provide } from 'vue';
import { GlobalContextParams } from './type';
import { configProviderInjectionKey, defaultContext } from './utils';

const props = withDefaults(defineProps<GlobalContextParams>(), defaultContext);
provide(configProviderInjectionKey, props);
</script>
<template>
    <slot />
</template>
