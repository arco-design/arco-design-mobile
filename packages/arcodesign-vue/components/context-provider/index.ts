import ContextProvider from './ContextProvider.vue';
import { installFactory } from '../_helpers/install';

export type * from './type';
export * from './utils';

const install = installFactory(ContextProvider);

export { ContextProvider };

export default {
    install,
};
