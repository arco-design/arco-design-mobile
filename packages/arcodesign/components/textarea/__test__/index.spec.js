import React from 'react';
import { mount } from 'enzyme';
import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import { defaultContext } from '../../context-provider';
import Textarea from '..';

const inputPrefix = `${defaultContext.prefixCls}-input`;

demoTest('textarea');

mountTest(Textarea, 'Textarea');

describe('Textarea', () => {
    it('should callback correctly when change text', () => {
        const mockFn = jest.fn();
        const component = mount(<Textarea onInput={mockFn}/>);
        component.find('textarea').simulate('input', {
            target: {
                value: '123'
            }
        });
        expect(mockFn).toBeCalled();
    });
    it('should render and callback correctly when more than limit', () => {
        const mockFn = jest.fn();
        const component = mount(<Textarea statisticsMaxlength={5} onErrStatusChange={mockFn}/>);
        component.find('textarea').simulate('input', {
            target: {
                value: '12345678'
            }
        });
        expect(mockFn).toBeCalled();
        expect(component.find('.statistic-text').hasClass('exceed')).toBe(true);
        expect(component.find('.statistic-text').text()).toBe("8/5");
        expect(component.find('textarea').text()).toBe('12345678');
    });
    it('should render correctly when set area limit', () => {
        const component = mount(<Textarea textareaStyle={{ height: 55 }}/>);
        expect(component.find('textarea').props().style.height).toBe(55);
    });
    it('should render correctly when text calls the area resize', () => {
        // const component = mount(<Textarea autosize/>);
        // const initHeight = component.find('textarea').get(0).style;
        // console.log(component.find('textarea').html());
    });
    it('should render correctly when set prefix', () => {
        const component = mount(<Textarea prefix="prefix"/>);
        expect(component.find(`.${inputPrefix}-prefix`).length).toBe(1);
        expect(component.find(`.${inputPrefix}-prefix`).text()).toBe('prefix');
    });
});
