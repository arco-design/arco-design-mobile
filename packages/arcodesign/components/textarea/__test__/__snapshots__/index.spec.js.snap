// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textarea demo test textarea demo: error-hint.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-input-container all-border-box "
    role="search"
  >
    <div
      class="arco-input-wrap textarea has-stat border-none pc"
    >
      <textarea
        class="arco-textarea"
        placeholder="Please enter the description of no more than 5 characters"
        rows="1"
        style="height: 0px;"
      />
      <span
        class="statistic-text"
      >
        0/5
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`textarea demo test textarea demo: fix-height.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-input-container all-border-box "
    role="search"
  >
    <div
      class="arco-input-wrap textarea  border-none pc"
    >
      <textarea
        class="arco-textarea"
        placeholder="Please enter the description of no less than 10 characters"
        style="height: 55px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`textarea demo test textarea demo: index.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-input-container all-border-box "
    role="search"
  >
    <div
      class="arco-input-wrap textarea has-stat border-none pc"
    >
      <textarea
        class="arco-textarea"
        placeholder="Please enter the description of no less than 10 characters"
        style="height: 0px;"
      />
      <span
        class="statistic-text"
      >
        0/100
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`textarea demo test textarea demo: label.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-input-container all-border-box "
    role="search"
  >
    <div
      class="arco-input-wrap textarea has-stat border-none pc prefix"
    >
      <div
        class="arco-input-prefix"
      >
        Message
      </div>
      <textarea
        class="arco-textarea"
        placeholder="Please enter the description of no less than 10 characters"
        rows="2"
        style="height: 0px;"
      />
      <span
        class="statistic-text"
      >
        0/50
      </span>
    </div>
  </div>
</DocumentFragment>
`;
